src/ray/core_worker/transport/concurrency_group_manager.h

```diff
@@ -68,6 +76,12 @@ class ConcurrencyGroupManager final {
   // The default concurrency group executor. It's nullptr if its max concurrency is 1.
   std::shared_ptr<ExecutorType> default_executor_ = nullptr;
 
+  // The function that can be used to initialize the executor.
+  std::function<std::function<void()>()> initializer_;
+
+  // A vector of optional functions that can be used to release the executors.
+  std::vector<std::optional<std::function<void()>>> executor_releasers_;
```

Comment:
why's it need to be a vector?

Reply:
we should call it in each thread.

Reply:
Each releaser holds the `gstate` for a thread.

```cpp
gstate = PyGILState_Ensure();
...
PyGILState_Release(gstate);
```

Reply:
oh I see, does the index map to the threads then?

Reply:
> does the index map to the threads then?

No, the index doesn't mean anything. Each callback function in the vector looks like:

```cpp
executor->Post([...]() { PyGILState_Release(gstate); });
```

so it has already been associated to a thread.

Reply:
Oh I see -- so this is because we assume the lifetime of the threads are all coupled (they are all destructed when the process exits)

Reply:
> so this is because we assume the lifetime of the threads are all coupled (they are all destructed when the process exits)

Yes. Is the assumption correct? The implementation fulfills the assumption.

Reply:
yes it is for now